### PR TITLE
feat(scheduler): use an worker pool for scheduler

### PR DESCRIPTION
**Pull request by peusebiu**
Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -85,24 +85,42 @@ func NewScheduler(logC log.Logger) *Scheduler {
...
 }
 
-const rateLimit = 5 * time.Second
+const (
+	rateLimit  = 5 * time.Second
+	numWorkers = 5
+)
+
+func (scheduler *Scheduler) taskWorker(jobs chan Task) {
+	for j := range jobs {
```
Comment (laurentiuNiculae) [inline pkg/scheduler/scheduler.go:94]: This for loop will run indefinitely until the channel jobs is closed?

Comment (peusebiu) [reply pkg/scheduler/scheduler.go:94]: yes

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -85,26 +85,55 @@ func NewScheduler(logC log.Logger) *Scheduler {
...
-				task := scheduler.getTask()
-				if task != nil {
-					if err := task.DoWork(); err != nil {
-						scheduler.log.Error().Err(err).Msg("error while executing task")
+				i := 0
+				for i < numWorkers {
+					task := scheduler.getTask()
+					if task != nil {
+						// push tasks into worker pool
+						scheduler.log.Debug().Msg("scheduler: pushing task into worker pool")
```
Comment (rchincha) [inline pkg/scheduler/scheduler.go:130]: This has the problem that we no longer get FIFO semantics.

Comment (andaaron) [reply pkg/scheduler/scheduler.go:130]: Please elaborate on this, what are you referring to?

Comment (peusebiu) [reply pkg/scheduler/scheduler.go:130]: is "enqueue task" better? or this is about implementation and not logging messaging?

**Review by rchincha — APPROVED** [review]
Note that when tasks are dequeued and run, we no longer get FIFO semantics.
If we need stronger guarantees, we will revisit.

**Pull request merged**